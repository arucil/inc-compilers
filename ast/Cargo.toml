[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
