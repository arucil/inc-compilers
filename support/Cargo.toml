[package]
name = "support"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
assert_cmd = "2.0.4"
predicates = "2.1.1"
pretty_assertions = "1.2.1"
scopeguard = "1.1.0"
