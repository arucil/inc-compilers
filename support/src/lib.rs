pub mod compile_error;
pub mod range;
pub mod test_cli;

pub use self::compile_error::*;
pub use self::range::*;
pub use self::test_cli::*;
