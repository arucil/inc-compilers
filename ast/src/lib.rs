#![feature(never_type, box_syntax)]

use id_arena::{Arena, Id};
use pretty::*;
use std::fmt::{self, Debug, Formatter};
use std::iter;
use support::Range;

pub mod parser;

pub use parser::{parse, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program<VAR = String, TYPE = ()> {
  pub body: Vec<Exp<VAR, TYPE>>,
  pub types: Arena<CompType>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Exp<VAR = String, TYPE = ()> {
  pub kind: ExpKind<VAR, TYPE>,
  pub range: Range,
  pub ty: TYPE,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExpKind<VAR, TYPE> {
  Int(i64),
  Prim {
    op: (Range, &'static str),
    args: Vec<Exp<VAR, TYPE>>,
  },
  Var(VAR),
  Str(String),
  Let {
    var: (Range, VAR),
    init: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Bool(bool),
  If {
    cond: Box<Exp<VAR, TYPE>>,
    conseq: Box<Exp<VAR, TYPE>>,
    alt: Box<Exp<VAR, TYPE>>,
  },
  Set {
    var: (Range, VAR),
    exp: Box<Exp<VAR, TYPE>>,
  },
  Get(VAR),
  Begin {
    seq: Vec<Exp<VAR, TYPE>>,
    last: Box<Exp<VAR, TYPE>>,
  },
  While {
    cond: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Void,
  Print(
    /// nonempty
    Vec<Exp<VAR, TYPE>>,
  ),
  NewLine,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Type {
  Int,
  Bool,
  Str,
  Comp(TypeId),
  Void,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TypeId(Id<CompType>);

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompType {
  Vector(Vec<Type>),
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct IdxVar {
  pub name: String,
  pub index: usize,
}

impl IdxVar {
  pub fn new<S: ToString>(name: S) -> Self {
    let name = name.to_string();
    if let Some(i) = name.find('.') {
      if let Ok(index) = name[i + 1..].parse::<usize>() {
        return Self {
          name: name[..i].to_owned(),
          index,
        };
      }
    }
    Self { name, index: 0 }
  }
}

impl Debug for IdxVar {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    write!(f, "{}.{}", self.name, self.index)
  }
}

impl<VAR: Debug, TYPE: Debug> Program<VAR, TYPE> {
  pub fn to_string_pretty(&self) -> String {
    let doc = self.to_doc();
    let mut buf = String::new();
    doc.render_fmt(30, &mut buf).unwrap();
    buf
  }

  fn to_doc(&self) -> RcDoc {
    RcDoc::intersperse(self.body.iter().map(|exp| exp.to_doc()), Doc::line())
  }
}

impl<VAR: Debug, TYPE: Debug> Exp<VAR, TYPE> {
  fn to_doc(&self) -> RcDoc {
    match &self.kind {
      ExpKind::Int(n) => RcDoc::text(format!("{}", n)),
      ExpKind::Str(s) => RcDoc::text(format!("{:?}", s)),
      ExpKind::Var(var) | ExpKind::Get(var) => {
        RcDoc::text(format!("{:?}", var))
      }
      ExpKind::Prim { op, args } => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text(op.1))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Let { var, init, body } => RcDoc::text("(")
        .append(
          RcDoc::text("let")
            .append(Doc::line())
            .append(
              RcDoc::text("[")
                .append(
                  RcDoc::text(format!("{:?}", var.1))
                    .append(Doc::line())
                    .append(init.to_doc())
                    .nest(1)
                    .group(),
                )
                .append(RcDoc::text("]")),
            )
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Bool(true) => RcDoc::text("#t"),
      ExpKind::Bool(false) => RcDoc::text("#f"),
      ExpKind::If { cond, conseq, alt } => RcDoc::text("(")
        .append(
          RcDoc::text("if")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(conseq.to_doc().append(Doc::line()).append(alt.to_doc()))
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Set { var, exp } => RcDoc::text("(")
        .append(
          RcDoc::text("set!")
            .append(Doc::line())
            .append(RcDoc::text(format!("{:?}", var.1)))
            .append(Doc::line())
            .append(exp.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Begin { seq, last } => RcDoc::text("(")
        .append(
          RcDoc::text("begin").append(Doc::line()).append(
            RcDoc::intersperse(
              seq
                .iter()
                .chain(std::iter::once(&**last))
                .map(|exp| exp.to_doc()),
              Doc::line(),
            )
            .nest(1)
            .group(),
          ),
        )
        .append(RcDoc::text(")")),
      ExpKind::While { cond, body } => RcDoc::text("(")
        .append(
          RcDoc::text("while")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Void => RcDoc::text("(void)"),
      ExpKind::Print(args) => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text("print"))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::NewLine => RcDoc::text("(newline)"),
    }
  }
}
