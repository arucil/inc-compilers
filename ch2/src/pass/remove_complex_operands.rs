use ast::{Exp, ExpKind, IdxVar, Program};
use support::Range;

pub fn remove_complex_operands<TYPE: Clone>(
  prog: Program<IdxVar, TYPE>,
) -> Program<IdxVar, TYPE> {
  let mut counter = 0;
  Program {
    body: prog
      .body
      .into_iter()
      .map(|exp| mon_exp(exp, &mut counter))
      .collect(),
    ..prog
  }
}

struct TmpVar<TYPE> {
  range: Range,
  name: IdxVar,
  /// May be not atomic.
  init: Exp<IdxVar, TYPE>,
}

/// process expressions that do not need to be atomic.
///
/// `mon` stands for Monadic Normal Form.
fn mon_exp<TYPE: Clone>(
  exp: Exp<IdxVar, TYPE>,
  counter: &mut usize,
) -> Exp<IdxVar, TYPE> {
  let range = exp.range;
  let ty = exp.ty.clone();
  match exp.kind {
    ExpKind::Int(..) => exp,
    ExpKind::Var(..) => exp,
    ExpKind::Let { var, init, body } => Exp {
      kind: ExpKind::Let {
        var,
        init: box mon_exp(*init, counter),
        body: box mon_exp(*body, counter),
      },
      range,
      ty,
    },
    ExpKind::Prim { op, args } => mon_prim(
      args,
      |args| Exp {
        kind: ExpKind::Prim { op, args },
        range,
        ty,
      },
      counter,
    ),
    // ch4
    ExpKind::If { cond, conseq, alt } => Exp {
      kind: ExpKind::If {
        cond: box mon_exp(*cond, counter),
        conseq: box mon_exp(*conseq, counter),
        alt: box mon_exp(*alt, counter),
      },
      range,
      ty,
    },
    ExpKind::Bool(..) => exp,
    // ch5
    ExpKind::Get(..) => exp,
    ExpKind::Set { var, exp } => Exp {
      kind: ExpKind::Set {
        var,
        exp: box mon_exp(*exp, counter),
      },
      range,
      ty,
    },
    ExpKind::Begin { seq, last } => Exp {
      kind: ExpKind::Begin {
        seq: seq.into_iter().map(|exp| mon_exp(exp, counter)).collect(),
        last: box mon_exp(*last, counter),
      },
      range,
      ty,
    },
    ExpKind::While { cond, body } => Exp {
      kind: ExpKind::While {
        cond: box mon_exp(*cond, counter),
        body: box mon_exp(*body, counter),
      },
      range,
      ty,
    },
    ExpKind::Void => exp,
    ExpKind::Print(args) => mon_prim(
      args,
      |args| Exp {
        kind: ExpKind::Print(args),
        range,
        ty,
      },
      counter,
    ),
    ExpKind::Str(..) => exp,
    ExpKind::NewLine => exp,
  }
}

fn mon_prim<TYPE: Clone>(
  args: Vec<Exp<IdxVar, TYPE>>,
  build_exp: impl FnOnce(Vec<Exp<IdxVar, TYPE>>) -> Exp<IdxVar, TYPE>,
  counter: &mut usize,
) -> Exp<IdxVar, TYPE> {
  let mut tmps = vec![];
  // TODO rearrange args to make atom exps come after complex ones, but make
  // sure their results in original order.
  let args = args
    .into_iter()
    .map(|arg| atom_exp(arg, &mut tmps, counter))
    .collect();
  tmps.into_iter().rfold(build_exp(args), |body, tmp| Exp {
    range: body.range,
    ty: body.ty.clone(),
    kind: ExpKind::Let {
      var: (tmp.range, tmp.name),
      init: box tmp.init,
      body: box body,
    },
  })
}

/// Process expressions that need to be atomic.
fn atom_exp<TYPE: Clone>(
  exp: Exp<IdxVar, TYPE>,
  tmps: &mut Vec<TmpVar<TYPE>>,
  counter: &mut usize,
) -> Exp<IdxVar, TYPE> {
  match exp.kind {
    ExpKind::Let { var, init, body } => {
      tmps.push(TmpVar {
        range: var.0,
        name: var.1,
        init: mon_exp(*init, counter),
      });
      atom_exp(*body, tmps, counter)
    }
    ExpKind::Prim { op, args } => {
      let args = args
        .into_iter()
        .map(|arg| atom_exp(arg, tmps, counter))
        .collect();
      let exp = Exp {
        kind: ExpKind::Prim { op, args },
        range: exp.range,
        ty: exp.ty,
      };
      assign_var(exp, tmps, counter)
    }
    ExpKind::Bool(..) | ExpKind::Int(..) | ExpKind::Var(..) => exp,
    // ch4
    ExpKind::If { .. } => {
      let exp = mon_exp(exp, counter);
      assign_var(exp, tmps, counter)
    }
    // ch5
    ExpKind::Get(..)
    | ExpKind::Begin { .. }
    | ExpKind::While { .. }
    | ExpKind::Set { .. }
    | ExpKind::Print { .. }
    | ExpKind::Str(..) => {
      let exp = mon_exp(exp, counter);
      assign_var(exp, tmps, counter)
    }
    ExpKind::Void => exp,
    // ExpKind::Str(..) => exp,
    ExpKind::NewLine => exp,
  }
}

fn assign_var<TYPE: Clone>(
  exp: Exp<IdxVar, TYPE>,
  tmps: &mut Vec<TmpVar<TYPE>>,
  counter: &mut usize,
) -> Exp<IdxVar, TYPE> {
  let range = exp.range;
  let ty = exp.ty.clone();
  let tmp = IdxVar {
    name: "tmp".to_owned(),
    index: *counter,
  };
  *counter += 1;
  tmps.push(TmpVar {
    range,
    name: tmp.clone(),
    init: exp,
  });
  Exp {
    kind: ExpKind::Var(tmp),
    range,
    ty,
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (- y (- 2))))"#)
        .unwrap();
    let prog = uniquify::uniquify(prog).unwrap();
    let result = remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn init_with_var() {
    let prog = parse(r#"(let ([a 42]) (let ([b a]) b))"#).unwrap();
    let prog = uniquify::uniquify(prog).unwrap();
    let result = remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_in_init() {
    let prog = parse(
      r#"
      (let
        ([a
          (+ (let
               ([x (let
                     ([x (read)])
                     (+ x 13))]
                [y (- x)])
               (+ x (- y)))
             7)])
        (- a))
      "#,
    )
    .unwrap();
    let prog = uniquify::uniquify(prog).unwrap();
    let result = remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
