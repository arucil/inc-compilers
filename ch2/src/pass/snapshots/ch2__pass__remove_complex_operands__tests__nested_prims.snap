---
source: ch2/src/pass/remove_complex_operands.rs
expression: result.to_string_pretty()
---
(let
 [x.0 (read)]
 (let
  [y.1 (+ 2 3)]
  (let
   [(tmp).0 (read)]
   (let
    [(tmp).1 (- (tmp).0)]
    (let
     [(tmp).2 (- 2)]
     (let
      [(tmp).3 (- y.1 (tmp).2)]
      (+ (tmp).1 (tmp).3)))))))
